src/metrics/readme_metrics.nit

```diff
@@ -0,0 +1,272 @@
+
+		var	metrics = new ReadmeMetrics
+		metrics.collect_metrics(model.mpackages)
+		metrics.to_console(toolcontext)
+
+		var csv = toolcontext.opt_csv.value
+		if csv then metrics.to_csv.write_to_file("{toolcontext.opt_dir.value or else "metrics"}/readme.csv")
+	end
+end
+
+# A Markdown decorator that collect metrics about a Readme content
+class MetricsDecorator
+	super HTMLDecorator
+
+	# Count blocks
+	var block_counter = new Counter[String]
+
+	# Count sections
+	var headline_counter = new Counter[Int]
+
+	redef fun add_ruler(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_headline(v, block) do
+		block_counter.inc block.class_name
+		headline_counter.inc block.depth
+		super
+	end
+
+	redef fun add_paragraph(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_code(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_blockquote(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_unorderedlist(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_orderedlist(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_listitem(v, block) do
+		block_counter.inc block.class_name
+		super
+	end
+
+	redef fun add_image(v, link, name, comment) do
+		block_counter.inc "Image"
+		super
+	end
+
+	redef fun add_link(v, link, name, comment) do
+		block_counter.inc "Link"
+		super
+	end
+
+	redef fun add_span_code(v, text, from, to) do
+		block_counter.inc "SpanCode"
+		super
+	end
+end
+
+# All metrics about the readmes
+class ReadmeMetrics
+	super HashMap[MPackage, ReadmeMetric]
+
+	# Collect all metric names from submetrics
+	fun metrics_names: ArraySet[String] do
+		var keys = new ArraySet[String]
+		keys.add "MPackage"
+		for mpackage, values in self do
+			keys.add_all values.keys
+		end
+		return keys
+	end
+
+	# Render `self` as a CsvDocument
+	fun to_csv: CsvDocument do
+		var doc = new CsvDocument
+		doc.header = metrics_names.to_a
+
+		var metrics = metrics_names
+		for mpackage in self.keys do
+			doc.records.add self[mpackage].to_csv_record(metrics)
+		end
+		return doc
+	end
+
+	# Print `self` into stdout
+	fun to_console(toolcontext: ToolContext) do
+		for mpackage, values in self do
+			if not values.has_readme then continue
+			values.to_console(toolcontext)
+		end
+	end
+
+	# Collect metrics for all `mpackages`
+	fun collect_metrics(mpackages: Collection[MPackage]) do
+		for mpackage in mpackages do
+			var metric = new ReadmeMetric(mpackage)
+			metric.collect_metrics
+			self[mpackage] = metric
+		end
+	end
+end
+
+# Readme metrics associated to a Package
+class ReadmeMetric
+	super HashMap[String, Int]
+
+	# Package this Readme is about
+	var mpackage: MPackage
+
+	# Render `self` as a CsvDocument record
+	fun to_csv_record(keys: ArraySet[String]): Array[String] do
+		var record = new Array[String]
+		record.add mpackage.full_name
+		for key in keys do
```

Comment:
maybe you want to sort the keys? by default it's the apparition order.

Reply:
I will let the CSV user do that.